lib/test/cdo/contact_rollups/test_pardot_v2.rb

```diff
@@ -30,69 +30,156 @@ def test_retrieve_new_ids_with_result
+    assert_equal expected_submissions, submissions
+    assert_equal expected_errors, errors
+  end
+
+  def test_batch_update_prospects_single_contact
+    # TODO: add more fields to new_contact_data so that new_contact_data and old_prospect_data
+    #   semantically overlap (field names could be different).
+    #   Same in test_batch_update_prospects_multiple_contacts.
+    contact = {
+      email: 'alpha@cdo.org',
+      pardot_id: 1,
+      old_prospect_data: {db_Opt_In: 'Yes'},
+      new_contact_data: {opt_in: false}
+    }
+
+    ok_response = Nokogiri.XML <<-XML
+      <rsp stat="ok" version="1.0">
+          <errors/>
+      </rsp>
+    XML
+    PardotV2.stubs(:post_with_auth_retry).once.returns(ok_response)
+
+    # Eagerly submit an update request
+    submissions, errors = PardotV2.new.batch_update_prospects(
+      *contact.values_at(:email, :pardot_id, :old_prospect_data, :new_contact_data),
+      true
+    )
+
+    expected_submissions = [
+      {email: contact[:email], id: contact[:pardot_id], db_Opt_In: 'No'}
+    ]
+    assert_equal expected_submissions, submissions
     assert_equal [], errors
+  end
 
-    # Second call, eagerly send request
-    submitted, errors = pardot_writer.batch_create_prospects(contacts.last, true)
+  def test_batch_update_prospects_multiple_contacts
+    contacts = [
+      {
+        email: 'alpha@cdo.org',
+        pardot_id: 1,
+        old_prospect_data: {db_Opt_In: 'Yes'},
+        new_contact_data: {opt_in: false}
+      },
+      {
+        email: 'beta@cdo.org',
+        pardot_id: 2,
+        old_prospect_data: nil,
+        new_contact_data: {opt_in: true}
+      }
+    ]
+
+    response_with_errors = Nokogiri.XML <<-XML
+      <rsp stat="fail" version="1.0">
+          <errors>
+              <prospect identifier="0">#{PardotHelpers::ERROR_INVALID_EMAIL}</prospect>
+          </errors>
+      </rsp>
+    XML
+    PardotV2.stubs(:post_with_auth_retry).once.returns(response_with_errors)
+
+    # Calling batch_update for each contact. No update request shall be sent
+    pardot_writer = PardotV2.new
+    contacts.each do |contact|
+      submissions, errors = pardot_writer.batch_update_prospects(
+        *contact.values_at(:email, :pardot_id, :old_prospect_data, :new_contact_data)
+      )
+      assert_equal [], submissions
+      assert_equal [], errors
+    end
+
+    # Flushing out the remaining data, request shall be sent immediately
+    submissions, errors = pardot_writer.batch_update_remaining_prospects
 
     expected_submissions = [
       {email: contacts.first[:email], id: contacts.first[:pardot_id], db_Opt_In: 'No'},
       {email: contacts.last[:email], id: contacts.last[:pardot_id], db_Opt_In: 'Yes'}
     ]
-    expected_errors = [{prospect_index: 0, error_msg: error_msg}]
-    assert_equal expected_submissions, submitted
+    expected_errors = [{prospect_index: 0, error_msg: PardotHelpers::ERROR_INVALID_EMAIL}]
+    assert_equal expected_submissions, submissions
```

Comment:
does submissions contain data for submissions that were rejected as well as those accepted?

Reply:
Yes, it contains both. `accepted` and `rejected` are better terms, I'll use them in function names.